### ISSUE TITLE
Add ZoomConstantExpression to StylePropertyExpression

### ISSUE THREAD

**Issue by tangerine-orange** [maplibre/maplibre-style-spec#267]
**Describe the bug**
In this piece of [code](https://github.com/maplibre/maplibre-gl-js/blob/main/src/style/style_layer/line_style_layer.ts#L63) from maplibre-gl-gs:
```
    _handleSpecialPaintPropertyUpdate(name: string) {
        if (name === 'line-gradient') {
            const expression: ZoomConstantExpression<'source'> = (this._transitionablePaint._values['line-gradient'].value.expression as any);
            this.stepInterpolant = expression._styleExpression.expression instanceof Step;
            this.gradientVersion = (this.gradientVersion + 1) % Number.MAX_SAFE_INTEGER;
        }
    }
```
we are coercing `expression` to `ZoomConstantExpression<'source'>`. This is required because `expression` is [defined](https://github.com/maplibre/maplibre-gl-js/blob/main/src/style/properties.ts#L71) as a `StylePropertyExpression = ConstantExpression | SourceExpression | CameraExpression | CompositeExpression`.

This coercion tells me that there's some problem with the typing here. I see two options:

1. expand `StylePropertyExpression` to include `ZoomConstantExpression`, and possibly `ZoomDependentExpression` as well. Currently, these are the only two expression types that `StylePropertyExpression` doesn't include.
2. In `mapbox-gl-gs` codebase, set `expression: StylePropertyExpression | ZoomConstantExpression<'source'>`

I'd think option 1 is better, but I'm very familiar with the types here. @birkskyum do you have an opinion?

---
This issue came up when investigating https://github.com/maplibre/maplibre-gl-js/pull/2837

**Comment by HarelM** [comment]
I think 1 is better, but we need to understand why it wasn't like that to begin with, i.e. is adding it to the type generalise it too much or is incorrect in some places...

**Comment by tangerine-orange** [comment]
agreed for sure

**Comment by birkskyum** [comment]
I don't know this area of the library well unfortunately.

**Comment by tangerine-orange** [comment]
Looking at this some more, I don't know if adding `ZoomConstantExpression` and `ZoomDependentExpression` to `StylePropertyExpression` is the right move. The structure of the zoom expressions is a bit different than the other style expressions. `expression`s are pretty complicated, with lots of nested expressions and different types of expressions.

Rather than just adding the zoom expressions to the `StylePropertyExpression` definition, a more comprehensive look at the typings of expressions is needed here.

For now, I think we should just expose an `isZoomExpression` type check like @HarelM suggests in https://github.com/maplibre/maplibre-gl-js/pull/2837.

**Comment by HarelM** [comment]
Fixed by #334
